pkg/core/resources.go

```diff
@@ -0,0 +1,32 @@
+package core
+
+import "github.com/dominikbraun/graph"
+
+type (
+	Construct interface {
+		Provenance() AnnotationKey
+		Id() string
+	}
+
+	CloudResource interface {
```

Comment:
Can we shorten this to `Resource`? Maybe even `Infrastructure` (if we're open to renaming). Since we're determining the new name and I wouldn't want to change it often, I think supporting on-prem and metal should indicate that this shouldn't be "cloud" specific (also why "provider" is named such).